Subject: [PATCH] 9_08_js_i18n
---
Index: src/main/webapp/resources/js/topjava.users.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/webapp/resources/js/topjava.users.js b/src/main/webapp/resources/js/topjava.users.js
--- a/src/main/webapp/resources/js/topjava.users.js	(revision 57fcaf20122970199f740c1e6e0e2f17e6202ed4)
+++ b/src/main/webapp/resources/js/topjava.users.js	(date 1744298386144)
@@ -14,10 +14,10 @@
     $.ajax({
         url: userAjaxUrl + id,
         type: "POST",
-        data: "enabled=" + enabled
+        data: `enabled=${enabled}`
     }).done(function () {
         chkbox.closest("tr").attr("data-user-enabled", enabled);
-        successNoty(enabled ? "Enabled" : "Disabled");
+        successNoty(enabled ? "common.enabled" : "common.disabled");
     }).fail(function () {
         $(chkbox).prop("checked", !enabled);
     });
@@ -41,7 +41,7 @@
                     "data": "email",
                     "render": function (data, type, row) {
                         if (type === "display") {
-                            return "<a href='mailto:" + data + "'>" + data + "</a>";
+                            return `<a href="mailto:${data}">${data}</a>`;
                         }
                         return data;
                     }
@@ -53,7 +53,7 @@
                     "data": "enabled",
                     "render": function (data, type, row) {
                         if (type === "display") {
-                            return "<input type='checkbox' " + (data ? "checked" : "") + " onclick='enable($(this)," + row.id + ");'/>";
+                            return `<input type='checkbox' ${data ? 'checked' : ''} onclick='enable($(this),${row.id});'/>`;
                         }
                         return data;
                     }
Index: src/main/webapp/resources/js/topjava.common.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/webapp/resources/js/topjava.common.js b/src/main/webapp/resources/js/topjava.common.js
--- a/src/main/webapp/resources/js/topjava.common.js	(revision 57fcaf20122970199f740c1e6e0e2f17e6202ed4)
+++ b/src/main/webapp/resources/js/topjava.common.js	(date 1744298438409)
@@ -13,28 +13,30 @@
 }
 
 function add() {
+    $("#modalTitle").html(i18n["addTitle"]);
     form.find(":input").val("");
     $("#editRow").modal();
 }
 
 function updateRow(id) {
     form.find(":input").val("");
+    $("#modalTitle").html(i18n["editTitle"]);
     $.get(ctx.ajaxUrl + id, function (data) {
         $.each(data, function (key, value) {
-            form.find("input[name='" + key + "']").val(value);
+            form.find(`input[name='${key}']`).val(value);
         });
         $('#editRow').modal();
     });
 }
 
 function deleteRow(id) {
-    if (confirm('Are you sure?')) {
+    if (confirm(i18n['common.confirm'])) {
         $.ajax({
             url: ctx.ajaxUrl + id,
             type: "DELETE"
         }).done(function () {
             ctx.updateTable();
-            successNoty("Deleted");
+            successNoty("common.deleted");
         });
     }
 }
@@ -51,7 +53,7 @@
     }).done(function () {
         $("#editRow").modal("hide");
         ctx.updateTable();
-        successNoty("Saved");
+        successNoty("common.saved");
     });
 }
 
@@ -64,10 +66,10 @@
     }
 }
 
-function successNoty(text) {
+function successNoty(key) {
     closeNoty();
     new Noty({
-        text: "<span class='fa fa-lg fa-check'></span> &nbsp;" + text,
+        text: `<span class='fa fa-lg fa-check'></span> &nbsp;${i18n[key]}`,
         type: 'success',
         layout: "bottomRight",
         timeout: 1000
@@ -76,20 +78,20 @@
 
 function renderEditBtn(data, type, row) {
     if (type === "display") {
-        return "<a onclick='updateRow(" + row.id + ");'><span class='fa fa-pencil'></span></a>";
+        return `<a onclick='updateRow(${row.id});'><span class='fa fa-pencil'></span></a>`;
     }
 }
 
 function renderDeleteBtn(data, type, row) {
     if (type === "display") {
-        return "<a onclick='deleteRow(" + row.id + ");'><span class='fa fa-remove'></span></a>";
+        return `<a onclick='deleteRow(${row.id});'><span class='fa fa-remove'></span></a>`;
     }
 }
 
 function failNoty(jqXHR) {
     closeNoty();
     failedNote = new Noty({
-        text: "<span class='fa fa-lg fa-exclamation-circle'></span> &nbsp;Error status: " + jqXHR.status + (jqXHR.responseJSON ? "<br>" + jqXHR.responseJSON : ""),
+        text: `<span class='fa fa-lg fa-exclamation-circle'></span> &nbsp;${i18n['common.errorStatus']}: ${jqXHR.status}${jqXHR.responseJSON ? '<br>' + jqXHR.responseJSON : ''}`,
         type: "error",
         layout: "bottomRight"
     });
Index: config/messages/app_ru.properties
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/config/messages/app_ru.properties b/config/messages/app_ru.properties
--- a/config/messages/app_ru.properties	(revision 57fcaf20122970199f740c1e6e0e2f17e6202ed4)
+++ b/config/messages/app_ru.properties	(date 1744296560878)
@@ -26,5 +26,11 @@
 
 common.add=Добавить
 common.select=Выбрать
+common.deleted=Запись удалена
+common.saved=Запись сохранена
+common.enabled=Запись активирована
+common.disabled=Запись деактивирована
+common.errorStatus=Статус ошибки
+common.confirm=Вы уверены?
 common.save=Сохранить
 common.cancel=Отменить
\ No newline at end of file
Index: config/messages/app.properties
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/config/messages/app.properties b/config/messages/app.properties
--- a/config/messages/app.properties	(revision 57fcaf20122970199f740c1e6e0e2f17e6202ed4)
+++ b/config/messages/app.properties	(date 1744296560867)
@@ -26,5 +26,11 @@
 
 common.add=Add
 common.select=Select
+common.deleted=Record deleted
+common.saved=Record saved
+common.enabled=Record enabled
+common.disabled=Record disabled
+common.errorStatus=Error status
+common.confirm=Are you sure?
 common.save=Save
 common.cancel=Cancel
\ No newline at end of file
Index: src/main/webapp/WEB-INF/jsp/users.jsp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/webapp/WEB-INF/jsp/users.jsp b/src/main/webapp/WEB-INF/jsp/users.jsp
--- a/src/main/webapp/WEB-INF/jsp/users.jsp	(revision 57fcaf20122970199f740c1e6e0e2f17e6202ed4)
+++ b/src/main/webapp/WEB-INF/jsp/users.jsp	(date 1744296560915)
@@ -37,7 +37,7 @@
     <div class="modal-dialog">
         <div class="modal-content">
             <div class="modal-header">
-                <h4 class="modal-title"><spring:message code="user.add"/></h4>
+                <h4 class="modal-title" id="modalTitle"></h4>
                 <button type="button" class="close" data-dismiss="modal" onclick="closeNoty()">&times;</button>
             </div>
             <div class="modal-body">
@@ -78,4 +78,13 @@
 </div>
 <jsp:include page="fragments/footer.jsp"/>
 </body>
+<script type="text/javascript">
+    const i18n = {}; // https://learn.javascript.ru/object
+    i18n["addTitle"] = '<spring:message code="user.add"/>';
+    i18n["editTitle"] = '<spring:message code="user.edit"/>';
+
+    <c:forEach var="key" items='${["common.deleted","common.saved","common.enabled","common.disabled","common.errorStatus","common.confirm"]}'>
+    i18n["${key}"] = "<spring:message code="${key}"/>";
+    </c:forEach>
+</script>
 </html>
\ No newline at end of file
